runtime/src/accounts_index.rs

```diff
@@ -70,33 +78,52 @@ impl<T: Clone> AccountsIndex<T> {
         account_info: T,
         reclaims: &mut Vec<(Fork, T)>,
     ) {
-        let roots = &self.roots;
-        let fork_vec = self
+        let _fork_vec = self
             .account_maps
             .entry(*pubkey)
-            .or_insert_with(|| (Vec::with_capacity(32)));
-
-        // filter out old entries
-        reclaims.extend(fork_vec.iter().filter(|(f, _)| *f == fork).cloned());
-        fork_vec.retain(|(f, _)| *f != fork);
-
-        // add the new entry
-        fork_vec.push((fork, account_info));
-
-        let max_root = Self::get_max_root(roots, fork_vec);
+            .or_insert_with(|| RwLock::new(Vec::with_capacity(32)));
+        self.update(fork, pubkey, account_info, reclaims);
+    }
 
-        reclaims.extend(
-            fork_vec
-                .iter()
-                .filter(|(fork, _)| Self::can_purge(max_root, *fork))
-                .cloned(),
-        );
-        fork_vec.retain(|(fork, _)| !Self::can_purge(max_root, *fork));
+    pub fn update(
+        &self,
+        fork: Fork,
+        pubkey: &Pubkey,
+        account_info: T,
+        reclaims: &mut Vec<(Fork, T)>,
+    ) -> Option<T> {
+        let roots = &self.roots;
+        if let Some(lock) = self.account_maps.get(pubkey) {
+            let mut fork_vec = lock.write().unwrap();
+            // filter out old entries
+            reclaims.extend(fork_vec.iter().filter(|(f, _)| *f == fork).cloned());
+            fork_vec.retain(|(f, _)| *f != fork);
+
+            // add the new entry
+            fork_vec.push((fork, account_info));
+
+            let max_root = Self::get_max_root(roots, &fork_vec);
+
+            reclaims.extend(
+                fork_vec
+                    .iter()
+                    .filter(|(fork, _)| Self::can_purge(max_root, *fork))
+                    .cloned(),
+            );
+            fork_vec.retain(|(fork, _)| !Self::can_purge(max_root, *fork));
+
+            return None;
+        } else {
+            return Some(account_info);
```

Comment:
what's this for?

Reply:
If it is Some(), means that we could not make the change without inserting a new value and thus needing a write lock on the accounts_index. If it is None, then it managed to do the update successfully.